fabric-resource-loader-v0/src/main/java/net/fabricmc/fabric/impl/resource/loader/ModResourcePackUtil.java

```diff
@@ -67,33 +73,58 @@ public static void appendModResourcePacks(List<ModResourcePack> packs, ResourceT
 				continue;
 			}
 
-			ModResourcePack pack = ModNioResourcePack.create(container.getMetadata().getId(), container, subPath, type, ResourcePackActivationType.ALWAYS_ENABLED);
+			ModResourcePack pack = ModNioResourcePack.create(container.getMetadata().getId(), container, subPath, type, ResourcePackActivationType.ALWAYS_ENABLED, true);
 
 			if (pack != null) {
 				packs.add(pack);
 			}
 		}
 	}
 
-	public static boolean containsDefault(ModMetadata info, String filename) {
-		return "pack.mcmeta".equals(filename);
+	public static boolean containsDefault(String filename, boolean modBundled) {
+		return "pack.mcmeta".equals(filename) || (modBundled && "pack.png".equals(filename));
 	}
 
-	public static InputStream openDefault(ModMetadata info, ResourceType type, String filename) {
+	public static InputStream getDefaultIcon() throws IOException {
+		Optional<Path> loaderIconPath = FabricLoader.getInstance().getModContainer("fabric-resource-loader-v0")
+				.flatMap(resourceLoaderContainer -> resourceLoaderContainer.getMetadata().getIconPath(512).flatMap(resourceLoaderContainer::findPath));
+
+		if (loaderIconPath.isPresent()) {
+			return Files.newInputStream(loaderIconPath.get());
+		}
+
+		// Should never happen in practice
+		return null;
+	}
+
+	public static InputStream openDefault(ModContainer container, ResourceType type, String filename) throws IOException {
 		switch (filename) {
 		case "pack.mcmeta":
-			String description = Objects.requireNonNullElse(info.getName(), "");
+			String description = Objects.requireNonNullElse(container.getMetadata().getName(), "");
 			String metadata = serializeMetadata(SharedConstants.getGameVersion().getResourceVersion(type), description);
 			return IOUtils.toInputStream(metadata, Charsets.UTF_8);
+		case "pack.png":
+			Optional<Path> path = container.getMetadata().getIconPath(512).flatMap(container::findPath);
+
+			if (path.isPresent()) {
+				return Files.newInputStream(path.get());
+			} else {
+				return getDefaultIcon();
+			}
 		default:
 			return null;
 		}
 	}
 
-	public static String serializeMetadata(int packVersion, String description) {
+	public static JsonObject getMetadataPackJson(int packVersion, JsonElement description) {
 		JsonObject pack = new JsonObject();
 		pack.addProperty("pack_format", packVersion);
-		pack.addProperty("description", description);
+		pack.add("description", description);
+		return pack;
+	}
+
+	public static String serializeMetadata(int packVersion, String description) {
+		JsonObject pack = getMetadataPackJson(packVersion, new JsonPrimitive(description));
```

Comment:
Might be nice for this to use the existing vanilla PackMetadata classes + codecs to create the json. This is something I changed in my PR here: https://github.com/FabricMC/fabric/pull/3312/files#diff-abb5e49de03dd72c0540388d57e30fcb1bbb255800703e9e69ce9e1e97aff21cR87

Not a blocker for this PR, just maybe nice to have.